cub/agent/agent_reduce.cuh

```diff
@@ -377,14 +377,17 @@ struct AgentReduce
     even_share.block_offset += even_share.block_stride;
 
     // Consume subsequent full tiles of input
-    while (even_share.block_offset + TILE_ITEMS <= even_share.block_end)
+    while (even_share.block_offset <= even_share.block_end - TILE_ITEMS)
```

Comment:
What happens if `even_share.block_end < TILE_ITEMS`? Won't the following check be better:

```
while (even_share.block_end - even_share.block_offset >= TILE_ITEMS)
```

Reply:
In my understanding, the only condition in which we get `even_share.block_end < TILE_ITEMS` is when there's one CTA per range. In this case, we should never reach this code and instead exists the function in the check above:
```cpp
if (even_share.block_end - even_share.block_offset < TILE_ITEMS)
{
  // First tile isn't full (not all threads have valid items)
  int valid_items = even_share.block_end - even_share.block_offset;
  ConsumeTile<true>(thread_aggregate,
                    even_share.block_offset,
                    valid_items,
                    Int2Type<false>(),
                    can_vectorize);
  return BlockReduceT(temp_storage.reduce)
    .Reduce(thread_aggregate, reduction_op, valid_items);
}
```
The motivation behind `while (even_share.block_offset <= even_share.block_end - TILE_ITEMS)` was that this way we can sum offsets (`even_share.block_end - TILE_ITEMS`) once per invocation, instead once per iteration in the `while (even_share.block_end - even_share.block_offset >= TILE_ITEMS)` case. 

Is there any other ways in which the proposed code is better?

Reply:
I think we would not run into the underflow case, where `even_share.block_end < TILE_ITEMS`, because we would return early by branching into this conditional (above):
```
if (even_share.block_end - even_share.block_offset < TILE_ITEMS)
```

Nonetheless, I think @canonizer 's suggestion would be preferable here.

Reply:
Benchmarked the difference between `while (even_share.block_offset <= even_share.block_end - TILE_ITEMS)` and `while (even_share.block_end - even_share.block_offset >= TILE_ITEMS)`. It seems that my motivation above is verified. Got up to 5% stable slowdown for `while (even_share.block_end - even_share.block_offset >= TILE_ITEMS)` on int16_t and uint32_t. Will go with `while (even_share.block_offset <= even_share.block_end - TILE_ITEMS)`.

Reply:
Interesting! Good to know. Thanks for checking.

Reply:
It looks like at this point, it holds that `even_share.block_end >= TILE_ITEMS`, so there won't be any underflow. If you want to keep the code as is, it's a good idea to add a comment.

I'd personally prefer `even_share.block_end - even_share.block_offset >= TILE_ITEMS` for uniformity, but it's up to you.

Reply:
> I'd personally prefer even_share.block_end - even_share.block_offset >= TILE_ITEMS for uniformity, but it's up to you.

I believe it's a matter of performance rather than preference. [Here](https://github.com/NVIDIA/cub/pull/592#discussion_r1027374025) I share 5% regression from this suggestion. And at the beginning of this thread is the intuition for the source of this regression.